spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala

```diff
@@ -296,3 +299,601 @@ class SampleToBatch[T: ClassTag]
     }
   }
 }
+
+/**
+ * Convert a sequence of [[Sample]] to a sequence of [[MiniBatch]] through function toMiniBatch.
+ *
+ * @param totalBatch total batch size
+ * @param partitionNum partition number of dataset, default means partitionNum
+ *                     equals Engine.nodeNumber()
+ * @param toMiniBatch toMiniBatch is an function convert an Array[Sample] to a MiniBatch[T], defined
+ *                    as (Array[Sample[T]], Array[Tensor[T]], Array[Tensor[T]]) => MiniBatch[T]).
+ *                    The two array[Tensor] are input buffers and target buffers, their lengths
+ *                    equal to the Sample's numFeature and numLabel.
+ */
+class SampleToMiniBatch[T: ClassTag](
+    totalBatch: Int,
+    partitionNum: Option[Int] = None,
+    toMiniBatch: (Array[Sample[T]], Array[Tensor[T]], Array[Tensor[T]]) => MiniBatch[T])
+    (implicit ev: TensorNumeric[T]) extends Transformer[Sample[T], MiniBatch[T]] {
+
+  private val batchPerPartition = Utils.getBatchSize(totalBatch)
+  private var inputBuffer: Array[Tensor[T]] = null
+  private var targetBuffer: Array[Tensor[T]] = null
+
+  override def apply(prev: Iterator[Sample[T]]): Iterator[MiniBatch[T]] = {
+    val batchSizePerPartition = batchPerPartition
+    new Iterator[MiniBatch[T]] {
+      private val batchSize = batchSizePerPartition
+
+      private val sampleData = new Array[Sample[T]](batchSize)
+      override def hasNext: Boolean = prev.hasNext
+
+      override def next(): MiniBatch[T] = {
+        if (prev.hasNext) {
+          var i = 0
+          while (i < batchSize && prev.hasNext) {
+            val sample = prev.next()
+            if (null == sampleData(i)) {
+              sampleData(i) = sample.clone()
+            } else {
+              sampleData(i).copy(sample)
+            }
+            i += 1
+          }
+          if (null == inputBuffer) {
+            inputBuffer = Array.tabulate(sampleData(0).numFeature())(_ => Tensor[T]())
+          }
+          if (null == targetBuffer && sampleData(0).numLabel() > 0) {
+            targetBuffer = Array.tabulate(sampleData(0).numLabel())(_ => Tensor[T]())
+          }
+
+          if (i == batchSize) {
+            toMiniBatch(sampleData, inputBuffer, targetBuffer)
+          } else {
+            // Deal with number Sample is smaller than batchSize.
+            toMiniBatch(sampleData.slice(0, i), inputBuffer, targetBuffer)
+          }
+        } else {
+          null
+        }
+      }
+    }
+  }
+}
+
+object SampleToMiniBatch {
+  /**
+   * Apply an SampleToMiniBatch transformer.
+   *
+   * @param batchSize total batch size
+   * @param toMiniBatch array[Sample] to MiniBatch function
+   * @return
+   */
+  def apply[T: ClassTag](
+        batchSize : Int,
+        toMiniBatch : (Array[Sample[T]], Array[Tensor[T]], Array[Tensor[T]]) => MiniBatch[T]
```

Comment:
what about move toMiniBatch into MiniBatch

And pass in a MiniBatch to the SampleToMiniBatch

We can have a default one.

MiniBatch.copy(samples: Seq[Sample[T]])